### PR TITLE
Add a new parameter for toggling jailed status of offline validator 

### PR DESCRIPTION
**Pull request by atchapcyp**
To easier create fake validators. This PR add a state parameter `JailOfflineValidator` to be a flag.

- add `set-jail-offline` command to set `JailOfflineValidator` flag.
- This flag is default `false` and require `dposv3.3` to be able to set by command.
- add message structure in `go-loom`

ref: #1303 
- [x] I added unit tests for any code that added
- [ ] I updated the CHANGELOG.md 
- [ ] All IP is original and not copied from another source
- [x] I assign all copyright to Loom Network for the code in the pull request

### CODE REVIEW

File: builtin/plugins/dposv3/dpos.go [hunk 1]
```diff
@@ -1336,6 +1337,29 @@ func (c *DPOS) ListAllDelegations(ctx contract.StaticContext, req *ListAllDelega
 	}, nil
 }
 
+func (c *DPOS) SetJailOfflineValidator(ctx contract.Context, req *SetJailOfflineValidatorRequest) error {
+	sender := ctx.Message().Sender
+	ctx.Logger().Info("DPOSv3 SetJailValidator", "request", req)
```
Comment (enlight) [inline builtin/plugins/dposv3/dpos.go]: You can remove this log line, it's not very useful

File: builtin/plugins/dposv3/dpos.go [hunk 2]
```diff
@@ -1336,6 +1337,29 @@ func (c *DPOS) ListAllDelegations(ctx contract.StaticContext, req *ListAllDelega
...
+	state, err := LoadState(ctx)
+	if err != nil {
+		return err
+	}
+
+	if state.Params.JailOfflineValidator == req.Jailed {
+		return nil
+	}
+
+	if state.Params.OracleAddress == nil || sender.Compare(loom.UnmarshalAddressPB(state.Params.OracleAddress)) != 0 {
```
Comment (enlight) [inline builtin/plugins/dposv3/dpos.go]: Do this before `if state.Params.JailOfflineValidator == req.Jailed`

File: builtin/plugins/dposv3/dpos.go [hunk 3]
```diff
@@ -1336,6 +1337,29 @@ func (c *DPOS) ListAllDelegations(ctx contract.StaticContext, req *ListAllDelega
 	}, nil
 }
 
+func (c *DPOS) SetJailOfflineValidator(ctx contract.Context, req *SetJailOfflineValidatorRequest) error {
```
Comment (enlight) [inline builtin/plugins/dposv3/dpos.go]: `SetJailOfflineValidator` -> `EnableValidatorJailing`

File: builtin/plugins/dposv3/dpos.go [hunk 4]
```diff
@@ -1336,6 +1337,29 @@ func (c *DPOS) ListAllDelegations(ctx contract.StaticContext, req *ListAllDelega
...
+	if !ctx.FeatureEnabled(loomchain.DPOSVersion3_3, false) {
+		return errors.New("DPOS v3.3 is not enabled")
+	}
+
+	state, err := LoadState(ctx)
+	if err != nil {
+		return err
+	}
+
+	if state.Params.JailOfflineValidator == req.Jailed {
```
Comment (enlight) [inline builtin/plugins/dposv3/dpos.go]: `state.Params.JailOfflineValidator` -> `state.Params.JailOfflineValidators`
`req.Jailed` -> `req.JailOfflineValidators`

File: cmd/loom/dposV3_commands.go [hunk 5]
```diff
@@ -916,6 +916,31 @@ func SetDowntimePeriodCmdV3() *cobra.Command {
 	return cmd
 }
 
+func SetJailOfflineValidatorCmdV3() *cobra.Command {
```
Comment (enlight) [inline cmd/loom/dposV3_commands.go]: -> `EnableValidatorJailingCmd()`
Can drop the V3 suffix.

File: cmd/loom/dposV3_commands.go [hunk 6]
```diff
@@ -916,6 +916,31 @@ func SetDowntimePeriodCmdV3() *cobra.Command {
...
+	var flags cli.ContractCallFlags
+	cmd := &cobra.Command{
+		Use:   "set-jail-offline [bool status] ",
```
Comment (enlight) [inline cmd/loom/dposV3_commands.go]: -> `enable-validator-jailing [enable]`

File: cmd/loom/dposV3_commands.go [hunk 7]
```diff
@@ -916,6 +916,31 @@ func SetDowntimePeriodCmdV3() *cobra.Command {
...
+		Short: "Set Jailed status to offline validator",
```
Comment (enlight) [inline cmd/loom/dposV3_commands.go]: -> `Toggle jailing of offline validators`

File: builtin/plugins/dposv3/dpos.go [hunk 8]
```diff
@@ -1382,7 +1404,7 @@ func UpdateDowntimeRecord(ctx contract.Context, downtimePeriod uint64, validator
 	)
 
 	// if DPOSv3.3 enabled, jail a valdiator that have been offline for last 4 periods
-	if ctx.FeatureEnabled(loomchain.DPOSVersion3_3, false) {
+	if ctx.FeatureEnabled(loomchain.DPOSVersion3_3, false) && enableJailOffline {
```
Comment (enlight) [inline builtin/plugins/dposv3/dpos.go]: This is getting a bit tricky, but it should probably look like this:
```golang
jailOfflineValidator := false
if ctx.FeatureEnabled(loomchain.DPOSVersion3_3, false) {
  jailOfflineValidator = true
}
if ctx.FeatureEnabled(loomchain.DPOSVersion3_5, false) {
  jailOfflineValidator = jailingEnabled
}

if jailOfflineValidator {
  // jail the validator
}
```

File: builtin/plugins/dposv3/dpos.go [hunk 9]
```diff
@@ -1368,7 +1390,7 @@ func ShiftDowntimeWindow(ctx contract.Context, currentHeight int64, candidates [
 	return nil
 }
 
-func UpdateDowntimeRecord(ctx contract.Context, downtimePeriod uint64, validatorAddr loom.Address) error {
+func UpdateDowntimeRecord(ctx contract.Context, downtimePeriod uint64, enableJailOffline bool, validatorAddr loom.Address) error {
```
Comment (enlight) [inline builtin/plugins/dposv3/dpos.go]: `enableJailOffline` -> `jailingEnabled`

File: cmd/loom/dposV3_commands.go [hunk 10]
```diff
@@ -916,12 +1046,47 @@ func SetDowntimePeriodCmdV3() *cobra.Command {
 	return cmd
 }
 
+const enebleValidatorJailingCmdExample = `
```
Comment (enlight) [inline cmd/loom/dposV3_commands.go]: -> `enableValidatorJailingCmdExample`

File: builtin/plugins/dposv3/test_helpers.go [hunk 11]
```diff
@@ -262,6 +262,14 @@ func (dpos *testDPOSContract) Unjail(ctx *plugin.FakeContext, candidate *loom.Ad
 	return err
 }
 
+func (dpos *testDPOSContract) EnableValidatorJailing(ctx *plugin.FakeContext, status bool) error {
+
```
Comment (enlight) [inline builtin/plugins/dposv3/test_helpers.go]: Remove the blank line

File: builtin/plugins/dposv3/dpos.go [hunk 12]
```diff
@@ -1336,6 +1337,27 @@ func (c *DPOS) ListAllDelegations(ctx contract.StaticContext, req *ListAllDelega
 	}, nil
 }
 
+func (c *DPOS) EnableValidatorJailing(ctx contract.Context, req *EnableValidatorJailingRequest) error {
+	sender := ctx.Message().Sender
+	if !ctx.FeatureEnabled(loomchain.DPOSVersion3_3, false) {
```
[the remark this answers is not in the document]
Comment (atchapcyp) [reply builtin/plugins/dposv3/dpos.go]: Should I just change to v3.4 or else?

[the remark this answers is not in the document]
Comment (enlight) [reply builtin/plugins/dposv3/dpos.go]: Yes, v3.4 should be fine

File: builtin/plugins/dposv3/dpos.go [hunk 13]
```diff
@@ -1336,6 +1337,27 @@ func (c *DPOS) ListAllDelegations(ctx contract.StaticContext, req *ListAllDelega
...
+	if !ctx.FeatureEnabled(loomchain.DPOSVersion3_4, false) {
+		return errors.New("DPOS v3.4 is not enabled")
+	}
+
+	state, err := LoadState(ctx)
+	if err != nil {
+		return err
+	}
+	if state.Params.OracleAddress == nil || sender.Compare(loom.UnmarshalAddressPB(state.Params.OracleAddress)) != 0 {
+		return errors.New("Invalid Oracle Address")
```
Comment (enlight) [inline builtin/plugins/dposv3/dpos.go]: Looks like the tests broke:
```
--- FAIL: TestChangeParams (0.00s)
[2019-07-12T08:37:30.367Z]     require.go:157: 
[2019-07-12T08:37:30.367Z]         	Error Trace:	dpos_test.go:189
[2019-07-12T08:37:30.367Z]         	Error:      	Not equal: 
[2019-07-12T08:37:30.367Z]         	            	expected: &errors.errorString{s:"Function can only be called with oracle address."}
[2019-07-12T08:37:30.367Z]         	            	actual  : &errors.errorString{s:"Invalid Oracle Address"}
[2019-07-12T08:37:30.367Z]         	            	
[2019-07-12T08:37:30.367Z]         	            	Diff:
[2019-07-12T08:37:30.367Z]         	            	--- Expected
[2019-07-12T08:37:30.367Z]         	            	+++ Actual
[2019-07-12T08:37:30.367Z]         	            	@@ -1,2 +1,2 @@
[2019-07-12T08:37:30.367Z]         	            	-(*errors.errorString)(Function can only be called with oracle address.)
[2019-07-12T08:37:30.367Z]         	            	+(*errors.errorString)(Invalid Oracle Address)
[2019-07-12T08:37:30.367Z]         	            	 
[2019-07-12T08:37:30.367Z]         	Test:       	TestChangeParams
```

Probably should just `return errOnlyOracle` instead here.

**Pull request merged**